hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestBasicReadWriteWithDifferentConnectionRegistries.java

```diff
@@ -0,0 +1,177 @@
+
+  @ClassRule
+  public static final HBaseClassTestRule CLASS_RULE =
+    HBaseClassTestRule.forClass(TestBasicReadWriteWithDifferentConnectionRegistries.class);
+
+  private static final Logger LOG =
+    LoggerFactory.getLogger(TestBasicReadWriteWithDifferentConnectionRegistries.class);
+
+  private static final HBaseTestingUtil UTIL = new HBaseTestingUtil();
+
+  public enum RegistryImpl {
+    ZK,
+    RPC,
+    ZK_URI,
+    RPC_URI
+  }
+
+  @Parameter
+  public RegistryImpl impl;
+
+  @Rule
+  public final TableNameTestRule name = new TableNameTestRule();
+
+  private byte[] FAMILY = Bytes.toBytes("family");
+
+  private Connection conn;
+
+  @Parameters(name = "{index}: impl={0}")
+  public static List<Object[]> data() {
+    List<Object[]> data = new ArrayList<Object[]>();
+    for (RegistryImpl impl : RegistryImpl.values()) {
+      data.add(new Object[] { impl });
+    }
+    return data;
+  }
+
+  @BeforeClass
+  public static void setUpBeforeClass() throws Exception {
+    UTIL.startMiniCluster();
+  }
+
+  @AfterClass
+  public static void tearDownAfterClass() throws Exception {
+    UTIL.shutdownMiniCluster();
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    switch (impl) {
+      case ZK: {
+        Configuration conf = HBaseConfiguration.create();
+        conf.setClass(HConstants.CLIENT_CONNECTION_REGISTRY_IMPL_CONF_KEY,
+          ZKConnectionRegistry.class, ConnectionRegistry.class);
+        String quorum = UTIL.getZkCluster().getAddress().toString();
+        String path = UTIL.getConfiguration().get(HConstants.ZOOKEEPER_ZNODE_PARENT);
+        conf.set(HConstants.CLIENT_ZOOKEEPER_QUORUM, quorum);
+        conf.set(HConstants.ZOOKEEPER_ZNODE_PARENT, path);
+        LOG.info("connect to cluster through zk quorum={} and parent={}", quorum, path);
+        conn = ConnectionFactory.createConnection(conf);
+        break;
+      }
+      case RPC: {
+        Configuration conf = HBaseConfiguration.create();
+        conf.setClass(HConstants.CLIENT_CONNECTION_REGISTRY_IMPL_CONF_KEY,
+          RpcConnectionRegistry.class, ConnectionRegistry.class);
+        String bootstrapServers =
+          UTIL.getMiniHBaseCluster().getMaster().getServerName().getAddress().toString();
+        conf.set(RpcConnectionRegistry.BOOTSTRAP_NODES, bootstrapServers);
+        LOG.info("connect to cluster through rpc bootstrap servers={}", bootstrapServers);
+        conn = ConnectionFactory.createConnection(conf);
+        break;
+      }
+      case ZK_URI: {
+        String quorum = UTIL.getZkCluster().getAddress().toString();
+        String path = UTIL.getConfiguration().get(HConstants.ZOOKEEPER_ZNODE_PARENT);
+        URI connectionUri = new URI("hbase+zk://" + quorum + path);
+        LOG.info("connect to cluster through connection url: {}", connectionUri);
+        conn = ConnectionFactory.createConnection(connectionUri);
+        break;
+      }
+      case RPC_URI: {
+        URI connectionUri = new URI("hbase+rpc://"
```

Comment:
IIRC minicluster always has a single master / ZK quorum host.

Does this code handle mulitple masters / ZK hosts ?

Reply:
For rpc I could append region server as well but for zookeeper, mini cluster will only have one zk server.

I do not think this is important for this test, our ZKConnectionRegistry and RpcConnectionRegistry are known to accept multiple servers, you just need to write it with the correct pattern.

Reply:
I could add a test in hbase-client module to make sure that we can decode the different parts of an URI correctly when constructing related ConnectionRegistries.

Reply:
Yes, my concern is whether the java URL parser can correctly handle multiple host:port entries.

Reply:
Got it. Will do.